include/pybind11/pytypes.h

```diff
@@ -936,11 +936,18 @@ class str : public object {
 public:
     PYBIND11_OBJECT_CVT(str, object, detail::PyUnicode_Check_Permissive, raw_str)
 
-    str(const char *c, size_t n)
-        : object(PyUnicode_FromStringAndSize(c, (ssize_t) n), stolen_t{}) {
+    str(const char *c, ssize_t n)
+        : object(PyUnicode_FromStringAndSize(c, n), stolen_t{}) {
         if (!m_ptr) pybind11_fail("Could not allocate string object!");
     }
 
+    template <typename T,
+              typename = typename std::enable_if<std::is_integral<T>::value>::type>
+    str(const char *c, T n) : str(c, ssize_t{n}) { }
```

Comment:
> I had to add a templated overload for all
integral types and then specialize specifically for ssize_t (the
base case) and size_t (to remove the -Wnarrowing warning for it,
for convenience/backcompat). Adding only overloads for ssize_t
and size_t would not work as py::bytes{"foo", 3} would result in
ambiguous overload resolution

Reply:
ah, saw that but didn't fully appreciate before, sorry, thanks.
hm, it seems very heavy-handed.
I don't feel too strongly in favor of this PR. I almost feel it's best to keep as-is, and let people decide if they want to simply cast themselves to avoid the warning, or use a helper that we could provide, e.g. `py::save_ssize_t()` which throws a `ValueError`, maybe we have that already?

Reply:
I'm not really sure what you suggest that the helper would do?

Reply:
Make the conversion safe, at runtime.
Maybe that's not what you had in mind? Is this PR just meant to suppress the `-Wnarrowing` warning? Even if there is actually a `ssize_t` overflow?

Reply:
My point was indeed just to suppress the warning: as noted at https://github.com/pybind/pybind11/pull/2692#discussion_r693384234, CPython itself will check the input and return NULL (which pybind11 already checks for) and set an error if there's an overflow.